File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/Colourable.java
package pw.switchcraft.plethora.gameplay.modules.glasses.objects;

import dan200.computercraft.api.lua.IArguments;
import dan200.computercraft.api.lua.LuaException;
import pw.switchcraft.plethora.api.method.BasicMethod;
import pw.switchcraft.plethora.api.method.FutureMethodResult;
import pw.switchcraft.plethora.api.method.IUnbakedContext;

import static pw.switchcraft.plethora.core.ContextHelpers.safeFromTarget;

/**
 * An object which can be coloured.
 */
public interface Colourable {
	int DEFAULT_COLOUR = 0xFFFFFFFF;

	int getColour();

	void setColour(int colour);

	BasicMethod<Colourable> GET_COLOUR = BasicMethod.of(
		"getColour", "function():number -- Get the colour for this object.",
		Colourable::getColour, false
	);
	BasicMethod<Colourable> GET_COLOR = BasicMethod.alias(GET_COLOUR, "getColor");
	static FutureMethodResult getColour(IUnbakedContext<Colourable> unbaked, IArguments args) throws LuaException {
		return FutureMethodResult.result(safeFromTarget(unbaked).getColour() & 0xFFFFFFFFL);
	}

	BasicMethod<Colourable> SET_COLOUR = BasicMethod.of(
		"setColour", "function(colour|r:int, [g:int, b:int], [alpha:int]):number -- Set the colour for this object.",
		Colourable::setColour, false
	);
	BasicMethod<Colourable> SET_COLOR = BasicMethod.alias(SET_COLOUR, "setColor");
	static FutureMethodResult setColour(IUnbakedContext<Colourable> unbaked, IArguments args) throws LuaException {
		Colourable object = safeFromTarget(unbaked);

		switch (args.count()) {
			case 1:
				object.setColour(args.getInt(0));
				break;

			case 3: {
				int r = args.getInt(0) & 0xFF;
				int g = args.getInt(1) & 0xFF;
				int b = args.getInt(2) & 0xFF;
				object.setColour((r << 24) | (g << 16) | (b << 8) | object.getColour() & 0xFF);
				break;
			}

			default:
			case 4: {
				int r = args.getInt(0) & 0xFF;
				int g = args.getInt(1) & 0xFF;
				int b = args.getInt(2) & 0xFF;
				int a = args.getInt(3) & 0xFF;
				object.setColour((r << 24) | (g << 16) | (b << 8) | a);
				break;
			}
		}

		return FutureMethodResult.empty();
	}

	BasicMethod<Colourable> GET_ALPHA = BasicMethod.of(
		"getAlpha", "function():number -- Get the alpha for this object.",
		Colourable::getAlpha, false
	);
	static FutureMethodResult getAlpha(IUnbakedContext<Colourable> unbaked, IArguments args) throws LuaException {
		return FutureMethodResult.result(safeFromTarget(unbaked).getColour() & 0xFF);
	}

	BasicMethod<Colourable> SET_ALPHA = BasicMethod.of(
		"setAlpha", "function(alpha:number):number -- Set the alpha for this object.",
		Colourable::setAlpha, false
	);
	static FutureMethodResult setAlpha(IUnbakedContext<Colourable> unbaked, IArguments args) throws LuaException {
		Colourable object = safeFromTarget(unbaked);
		object.setColour((object.getColour() & ~0xFF) | (args.getInt(0) & 0xFF));
		return FutureMethodResult.empty();
	}
}

File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/Colourable.kt
package pw.switchcraft.plethora.gameplay.modules.glasses.objects

import dan200.computercraft.api.lua.IArguments
import pw.switchcraft.plethora.api.method.BasicMethod
import pw.switchcraft.plethora.api.method.FutureMethodResult
import pw.switchcraft.plethora.api.method.IUnbakedContext
import pw.switchcraft.plethora.core.ContextHelpers.safeFromTarget

const val DEFAULT_COLOUR = 0xFFFFFFFFL

/**
 * An object which can be coloured.
 */
interface Colourable {
  var colour: Int

  companion object {
    val GET_COLOUR = BasicMethod.of(
      "getColour", "function():number -- Get the colour for this object.",
      { unbaked, _ -> getColour(unbaked) }, false
    )
    val GET_COLOR = BasicMethod.alias(GET_COLOUR, "getColor")
    private fun getColour(unbaked: IUnbakedContext<Colourable>) =
      FutureMethodResult.result(safeFromTarget(unbaked).colour.toLong() and 0xFFFFFFFFL)

    val SET_COLOUR = BasicMethod.of(
      "setColour", "function(colour|r:int, [g:int, b:int], [alpha:int]):number -- Set the colour for this object.",
      { unbaked, args -> setColour(unbaked, args) }, false
    )
    val SET_COLOR = BasicMethod.alias(SET_COLOUR, "setColor")
    private fun setColour(unbaked: IUnbakedContext<Colourable>?, args: IArguments): FutureMethodResult {
      val obj = safeFromTarget(unbaked)
      when (args.count()) {
        1 -> obj.colour = args.getInt(0)
        3 -> {
          val r = args.getInt(0) and 0xFF
          val g = args.getInt(1) and 0xFF
          val b = args.getInt(2) and 0xFF
          obj.colour = r shl 24 or (g shl 16) or (b shl 8) or (obj.colour and 0xFF)
        }
        else -> {
          val r = args.getInt(0) and 0xFF
          val g = args.getInt(1) and 0xFF
          val b = args.getInt(2) and 0xFF
          val a = args.getInt(3) and 0xFF
          obj.colour = r shl 24 or (g shl 16) or (b shl 8) or a
        }
      }
      return FutureMethodResult.empty()
    }

    val GET_ALPHA = BasicMethod.of(
      "getAlpha", "function():number -- Get the alpha for this object.",
      { unbaked, _ -> getAlpha(unbaked) }, false
    )
    private fun getAlpha(unbaked: IUnbakedContext<Colourable>) =
      FutureMethodResult.result(safeFromTarget(unbaked).colour and 0xFF)

    val SET_ALPHA = BasicMethod.of(
      "setAlpha", "function(alpha:number):number -- Set the alpha for this object.",
      { unbaked, args -> setAlpha(unbaked, args) }, false
    )
    private fun setAlpha(unbaked: IUnbakedContext<Colourable>, args: IArguments): FutureMethodResult {
      val obj = safeFromTarget(unbaked)
      obj.colour = obj.colour and 0xFF.inv() or (args.getInt(0) and 0xFF)
      return FutureMethodResult.empty()
    }
  }
}
